### PR TITLE
GH-37381: [Python][CI][Packaging] Enable ORC in Windows wheels and Appveyor CI

### PR DESCRIPTION
**Pull request by wgtmac**
### Rationale for this change

The pyarrow orc reader always crashes when it tries to create an internal orc reader. This is caused by failing to read tz database on the local host. This also disables windows wheel build when ARROW_ORC is turned on.

### What changes are included in this PR?

Download IANA timezone database on the test host and explicitly setting TZDIR to make the CIs happy.

### Are these changes tested?

Make sure all python wheel windows CIs pass.

### Are there any user-facing changes?

No.
* GitHub Issue: #36026
* GitHub Issue: #33499
* GitHub Issue: #37381

### CODE REVIEW

**Comment by wgtmac** [comment]
@github-actions crossbow submit wheel-windows*

**Comment by wgtmac** [comment]
I tried to copy what you did in https://github.com/apache/arrow/pull/37382 but it failed to find orc. Am I missing something? Sorry to bother you. @raulcd

**Comment by kou** [comment]
Could you try this temporally?

```diff
diff --git a/dev/tasks/python-wheels/github.windows.yml b/dev/tasks/python-wheels/github.windows.yml
index 01f4977a9b..042a2f7f2f 100644
--- a/dev/tasks/python-wheels/github.windows.yml
+++ b/dev/tasks/python-wheels/github.windows.yml
@@ -56,7 +56,7 @@ jobs:
           if errorlevel 1 (
             archery docker build --no-pull python-wheel-windows-vs2019 || exit /B 1
           )
-          archery docker run --no-build -e SETUPTOOLS_SCM_PRETEND_VERSION={{ arrow.no_rc_version }} python-wheel-windows-vs2019
+          archery docker run -e SETUPTOOLS_SCM_PRETEND_VERSION={{ arrow.no_rc_version }} python-wheel-windows-vs2019
 
       - uses: actions/upload-artifact@v3
         with:
```

**Comment by wgtmac** [comment]
@github-actions crossbow submit wheel-windows*

**Comment by wgtmac** [comment]
@github-actions crossbow submit wheel-windows*

**Comment by wgtmac** [comment]
@github-actions crossbow submit wheel-windows*

**Comment by wgtmac** [comment]
@github-actions crossbow submit wheel-windows*

**Comment by wgtmac** [comment]
@github-actions crossbow submit wheel-windows*

**Comment by wgtmac** [comment]
@github-actions crossbow submit wheel-windows*

**Comment by wgtmac** [comment]
@github-actions crossbow submit wheel-windows*

**Comment by wgtmac** [comment]
```
C:\>curl https://cygwin.osuosl.org/noarch/release/tzdata/tzdata-2024a-1.tar.xz --output tzdata.tar.gz   || exit /B 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
 51  451k   51  234k    0     0   234k      0  0:00:01 --:--:--  0:00:01  300k
100  451k  100  451k    0     0   451k      0  0:00:01 --:--:--  0:00:01  466k
C:\>mkdir C:\Users\ContainerAdministrator\Downloads\test\tzdata 
C:\>tar --extract --file tzdata.tar.gz --directory C:\Users\ContainerAdministrator\Downloads\test\tzdata 
tar: Error opening archive: Can't initialize filter; unable to run program "xz -d -qq"
```

Do you know any approach to unzip tar.xz file on Windows in the CI build? @kou

**Comment by raulcd** [comment]
I would try installing [archiver](https://community.chocolatey.org/packages/archiver) from choco on the test container: https://github.com/apache/arrow/blob/main/ci/docker/python-wheel-windows-test-vs2019.dockerfile

Similar to how we install things with choco here: https://github.com/apache/arrow/blob/main/ci/docker/python-wheel-windows-vs2019.dockerfile#L27-L28

**Comment by kou** [comment]
archiver will work. 7zip will work too.

**Comment by wgtmac** [comment]
Let me try again. Thank you very much! @raulcd @kou

**Comment by wgtmac** [comment]
@github-actions crossbow submit wheel-windows*

**Comment by raulcd** [comment]
Amazing :partying_face: ! Great work @wgtmac !

File: ci/scripts/python_wheel_windows_test.bat [hunk 1]
```diff
@@ -56,8 +56,15 @@ python -c "import pyarrow.dataset" || exit /B 1
 python -c "import pyarrow.flight" || exit /B 1
 python -c "import pyarrow.fs" || exit /B 1
 python -c "import pyarrow.json" || exit /B 1
+python -c "import pyarrow.orc" || exit /B 1
 python -c "import pyarrow.parquet" || exit /B 1
 python -c "import pyarrow.substrait" || exit /B 1
 
+@rem Download IANA Timezone Database for ORC C++
```
Comment (pitrou) [inline ci/scripts/python_wheel_windows_test.bat:63]: So ORC needs this unconditionally? Why is that?

Comment (wgtmac) [reply ci/scripts/python_wheel_windows_test.bat:63]: The root cause is that when creating an ORC reader it tries to read the local timezone: https://github.com/apache/orc/blob/main/c%2B%2B/src/Reader.cc#L251 which by default searchs `/usr/share/zoneinfo` if env `TZDIR` is unset: https://github.com/apache/orc/blob/main/c%2B%2B/src/Timezone.cc#L659

So the workaround here is to download tzdb and set env `TZDIR` to make the test happy. I think we can improve this on the ORC side:
- More search path candidates for IANA TZDB.
- Read the TZDB lazily as the metadata field is optional and for timestamp type only: https://github.com/apache/orc-format/blob/main/src/main/proto/orc/proto/orc_proto.proto#L181
- Ship TZDB if the build process happens on Windows.

File: dev/tasks/python-wheels/github.windows.yml [hunk 2]
```diff
@@ -56,7 +56,7 @@ jobs:
           if errorlevel 1 (
             archery docker build --no-pull python-wheel-windows-vs2019 || exit /B 1
           )
-          archery docker run --no-build -e SETUPTOOLS_SCM_PRETEND_VERSION={{ arrow.no_rc_version }} python-wheel-windows-vs2019
+          archery docker run -e SETUPTOOLS_SCM_PRETEND_VERSION={{ arrow.no_rc_version }} python-wheel-windows-vs2019
```
Comment (wgtmac) [inline dev/tasks/python-wheels/github.windows.yml:59]: Should we revert this line and update the pre-built docker image instead? BTW, this is ready for review now. @raulcd @kou

Comment (kou) [reply dev/tasks/python-wheels/github.windows.yml:59]: Yes. Please revert this for now.
Could you update https://github.com/apache/arrow/blob/4cce72033cae666d22110ea4b8f32a11915214d6/.env#L97-L101 instead?

**Review by kou — APPROVED** [review]
+1

File: dev/tasks/python-wheels/github.windows.yml (hunk 2 above)
Comment (wgtmac) [reply dev/tasks/python-wheels/github.windows.yml:59]: What value should I set? `PYTHON_WHEEL_WINDOWS_IMAGE_REVISION=2024-03-19` ?

**Comment by wgtmac** [comment]
@github-actions crossbow submit wheel-windows*

File: dev/tasks/python-wheels/github.windows.yml (hunk 2 above)
Comment (kou) [reply dev/tasks/python-wheels/github.windows.yml]: Yes. We can use today to avoid conflicting existing revision.

**Comment by wgtmac** [comment]
> Revision: [1a2634b](https://github.com/apache/arrow/commit/1a2634b37723074f2b1cbb0765666981fb997575)
> 
> Submitted crossbow builds: [ursacomputing/crossbow @ actions-2e59d7101d](https://github.com/ursacomputing/crossbow/branches/all?query=actions-2e59d7101d)
> 
> Task	Status
> wheel-windows-cp310-amd64	[GitHub Actions](https://github.com/ursacomputing/crossbow/actions/runs/8353185177/job/22864501261)
> wheel-windows-cp311-amd64	[GitHub Actions](https://github.com/ursacomputing/crossbow/actions/runs/8353185202/job/22864501363)
> wheel-windows-cp312-amd64	[GitHub Actions](https://github.com/ursacomputing/crossbow/actions/runs/8353185084/job/22864501166)
> wheel-windows-cp38-amd64	[GitHub Actions](https://github.com/ursacomputing/crossbow/actions/runs/8353185160/job/22864501245)
> wheel-windows-cp39-amd64	[GitHub Actions](https://github.com/ursacomputing/crossbow/actions/runs/8353185176/job/22864501282)

```
Step 6/8 : RUN choco install -r -y --no-progress python --version=%PYTHON_VERSION%
 ---> Running in 99f097760eef
Installing the following packages:
python
By installing you accept licenses for the packages.
python not installed. An error occurred during installation:
 The remote server returned an error: (504) Gateway Timeout. Gateway Time-out
python package files install completed. Performing other installation steps.
The install of python was NOT successful.
python not installed. An error occurred during installation:
 The remote server returned an error: (504) Gateway Timeout. Gateway Time-out

Chocolatey installed 0/1 packages. 1 packages failed.
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).

Failures
 - python (exited 1) - python not installed. An error occurred during installation:
 The remote server returned an error: (504) Gateway Timeout. Gateway Time-out
The command 'cmd /S /C choco install -r -y --no-progress python --version=%PYTHON_VERSION%' returned a non-zero code: 1
Error: docker build --build-arg BUILDKIT_INLINE_CACHE=1 --build-arg python=3.9 -f D:\a\crossbow\crossbow\arrow\ci/docker/python-wheel-windows-test-vs2019.dockerfile -t ghcr.io/ursacomputing/arrow:python-3.9-wheel-windows-test-vs2019-2024-03-19 D:\a\crossbow\crossbow\arrow exited with non-zero exit code 1
docker pull ghcr.io/ursacomputing/arrow:python-3.9-wheel-windows-test-vs2019-2024-03-19 exited with non-zero exit code 1
Error: Process completed with exit code 1.
```

Failed with above logs. I don't think we need to rerun them.

**Comment by wgtmac** [comment]
Failure in [AMD64 Ubuntu 20.04 R 4.3 Force-Tests true](https://github.com/apache/arrow/actions/runs/8353163894/job/22867379973#logs) is also unrelated.

```
══ Failed tests ════════════════════════════════════════════════════════════════
── Error ('test-duckdb.R:286:5'): to_duckdb passing a connection ───────────────
Error in `unclass(unclass(table_four)$lazy_query$x)$table`: $ operator is invalid for atomic vectors

[ FAIL 1 | WARN 0 | SKIP 22 | PASS 8265 ]
Error: Test failures
Execution halted

1 error ✖ | 0 warnings ✔ | 2 notes ✖
1
Error: `docker-compose --file /home/runner/work/arrow/arrow/docker-compose.yml run --rm -e TZ=MART -e ARROW_R_FORCE_TESTS=true ubuntu-r` exited with a non-zero exit code 1, see the process log above.
```

**Comment by wgtmac** [comment]
@kou Do you want to take another pass?

**Comment by pitrou** [comment]
Right. It seems that the ORC project should first make it possible to use without the timezone data files.

**Comment by wgtmac** [comment]
This PR aims to enable ORC in the wheels CI in case there is any new regression. I will work on the ORC side to make it robust as much as possible.

**Comment by jorisvandenbossche** [comment]
For CI we also have Appveyor, which was enabled here as well. But the wheels is what we deliver to our users. If we enable ORC there, I think we need to make this more robust on our side on the short term (or disable it again in the wheels until this is fixed in ORC and we use that version)

**Comment by wgtmac** [comment]
> But the wheels is what we deliver to our users.

Sorry I'm not familiar with that. It sounds that we need to revert this change at least for wheels.
